ReactRouterOutletElement.tsx breaks build of Flow application using vaadin-router

### Description of the bug

If the Flow application is configured to use `vaadin-router` instead of `react-router` (e.g. by setting `reactEnable=false` in vaadin-maven-plugin), the production build fails on Vite execution because `react-router-dom` import cannot be resolved.

```
[ERROR] [vite]: Rollup failed to resolve import "react-router-dom" from "/xxxxx/vaadin/starters/skeleton-starter-flow-spring/src/main/frontend/generated/jar-resources/ReactRouterOutletElement.tsx".
[ERROR] This is most likely unintended because it can break your application at runtime.
[ERROR] If you do want to externalize this module explicitly add it to
```

The workaround is to exclude the `flow-react` dependency in the POM file.

If it is expected that `flow-react` should be excluded when using `vaadin-router`, it should be explicitly documented. In the [upgrade guide](https://vaadin.com/docs/latest/upgrading#hilla-react-dependencies) it looks like an optional step.

### Expected behavior

A Flow application using `vaadin-router` should not fail during build even if `flow-react` dependency is present.

### Minimal reproducible example

- Clone https://github.com/vaadin/skeleton-starter-flow-spring/
- Update `vaadin-maven-plugin` configuration in POM file adding `<reactEnable>false</reactEnable>`
- Do a production build `mvn clean package -Pproduction`
- Check Vite build error in the console

### Versions

- Vaadin / Flow version: 24.5-SNAPSHOT
- Java version: 17

--- a/flow-plugins/flow-plugin-base/src/main/java/com/vaadin/flow/plugin/base/BuildFrontendUtil.java
+++ b/flow-plugins/flow-plugin-base/src/main/java/com/vaadin/flow/plugin/base/BuildFrontendUtil.java
@@ -579,7 +579,8 @@ public class BuildFrontendUtil {
         }
 
         FrontendDependenciesScanner scanner = new FrontendDependenciesScanner.FrontendDependenciesScannerFactory()
-                .createScanner(false, adapter.getClassFinder(), true, null);
+                .createScanner(false, adapter.getClassFinder(), true, null,
+                        adapter.isReactEnabled());
         List<Product> commercialComponents = findCommercialFrontendComponents(
                 scanner, statsJsonContent);
         commercialComponents.addAll(findCommercialJavaComponents(adapter));

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/NodeTasks.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/NodeTasks.java
@@ -113,8 +113,6 @@ public class NodeTasks implements FallibleCommand {
 
         Set<String> webComponentTags = new HashSet<>();
 
-        final FeatureFlags featureFlags = options.getFeatureFlags();
-
         if (options.isFrontendHotdeploy()) {
             UsageStatistics.markAsUsed("flow/hotdeploy", null);
         }
@@ -122,9 +120,7 @@ public class NodeTasks implements FallibleCommand {
         if (options.isEnablePackagesUpdate() || options.isEnableImportsUpdate()
                 || options.isEnableConfigUpdate()) {
             frontendDependencies = new FrontendDependenciesScanner.FrontendDependenciesScannerFactory()
-                    .createScanner(!options.isUseByteCodeScanner(), classFinder,
-                            options.isGenerateEmbeddableWebComponents(),
-                            featureFlags);
+                    .createScanner(options);
 
             if (options.isProductionMode()) {
                 boolean needBuild = BundleValidationUtil.needsBuild(options,

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/Options.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/Options.java
@@ -707,7 +707,12 @@ public class Options implements Serializable {
         return this;
     }
 
-    protected FeatureFlags getFeatureFlags() {
+    /**
+     * Get the available feature flags.
+     *
+     * @return FeatureFlags object
+     */
+    public FeatureFlags getFeatureFlags() {
         if (featureFlags == null) {
             featureFlags = new FeatureFlags(lookup);
             if (javaResourceFolder != null) {

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/scanner/FrontendDependencies.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/scanner/FrontendDependencies.java
@@ -143,6 +143,12 @@ public class FrontendDependencies extends AbstractDependenciesScanner {
     public FrontendDependencies(ClassFinder finder,
             boolean generateEmbeddableWebComponents,
             FeatureFlags featureFlags) {
+        this(finder, generateEmbeddableWebComponents, null, true);
+    }
+
+    public FrontendDependencies(ClassFinder finder,
+            boolean generateEmbeddableWebComponents, FeatureFlags featureFlags,
+            boolean reactEnabled) {
         super(finder, featureFlags);
         log().info(
                 "Scanning classes to find frontend configurations and dependencies...");
@@ -162,7 +168,9 @@ public class FrontendDependencies extends AbstractDependenciesScanner {
                     visitEntryPoint(entryPoints.get(themeClass.getName()));
                 }
             }
-            computeReactClasses(finder);
+            if (reactEnabled) {
+                computeReactClasses(finder);
+            }
             computePackages();
             computePwaConfiguration();
             aggregateEntryPointInformation();

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/scanner/FrontendDependenciesScanner.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/scanner/FrontendDependenciesScanner.java
@@ -22,6 +22,7 @@ import java.util.Set;
 
 import com.vaadin.experimental.FeatureFlags;
 import com.vaadin.flow.server.PwaConfiguration;
+import com.vaadin.flow.server.frontend.FrontendUtils;
 import com.vaadin.flow.server.frontend.Options;
 import com.vaadin.flow.theme.AbstractTheme;
 import com.vaadin.flow.theme.ThemeDefinition;
@@ -108,10 +109,12 @@ public interface FrontendDependenciesScanner extends Serializable {
         }
 
         public FrontendDependenciesScanner createScanner(Options options) {
+            boolean reactEnabled = options.isReactEnabled() && FrontendUtils
+                    .isReactRouterRequired(options.getFrontendDirectory());
             return createScanner(!options.isUseByteCodeScanner(),
                     options.getClassFinder(),
                     options.isGenerateEmbeddableWebComponents(),
-                    options.getFeatureFlags(), options.isReactEnabled());
+                    options.getFeatureFlags(), reactEnabled);
         }
     }
 

--- a/flow-server/src/main/java/com/vaadin/flow/server/frontend/scanner/FrontendDependenciesScanner.java
+++ b/flow-server/src/main/java/com/vaadin/flow/server/frontend/scanner/FrontendDependenciesScanner.java
@@ -22,6 +22,7 @@ import java.util.Set;
 
 import com.vaadin.experimental.FeatureFlags;
 import com.vaadin.flow.server.PwaConfiguration;
+import com.vaadin.flow.server.frontend.Options;
 import com.vaadin.flow.theme.AbstractTheme;
 import com.vaadin.flow.theme.ThemeDefinition;
 
@@ -87,14 +88,30 @@ public interface FrontendDependenciesScanner extends Serializable {
                 boolean allDependenciesScan, ClassFinder finder,
                 boolean generateEmbeddableWebComponents,
                 FeatureFlags featureFlags) {
+            return createScanner(allDependenciesScan, finder,
+                    generateEmbeddableWebComponents, featureFlags, true);
+        }
+
+        public FrontendDependenciesScanner createScanner(
+                boolean allDependenciesScan, ClassFinder finder,
+                boolean generateEmbeddableWebComponents,
+                FeatureFlags featureFlags, boolean reactEnabled) {
             if (allDependenciesScan) {
                 // this dep scanner can't distinguish embeddable web component
                 // frontend related annotations
                 return new FullDependenciesScanner(finder, featureFlags);
             } else {
                 return new FrontendDependencies(finder,
-                        generateEmbeddableWebComponents, featureFlags);
+                        generateEmbeddableWebComponents, featureFlags,
+                        reactEnabled);
             }
+        }
+
+        public FrontendDependenciesScanner createScanner(Options options) {
+            return createScanner(!options.isUseByteCodeScanner(),
+                    options.getClassFinder(),
+                    options.isGenerateEmbeddableWebComponents(),
+                    options.getFeatureFlags(), options.isReactEnabled());
         }
     }
 

